### ISSUE TITLE
Implement RFC6275 for IPv6 and Docker Swarm

### ISSUE THREAD

**Issue by agowa338** [moby/moby#38748]
To make IPv6 routing with distributed docker swarm nodes easier, RFC6275 could be implemented.

A container than gets assigned a care-of address from the host it is running on and a home address from the swarm.
The home address is used to connect to the container, where as the care-of address is used by the swarm to communicate with the container. Therefore allowing to distribute swarm nodes within multiple providers/data centers and moving containers around as needed.

The RFC also defines anti-replay and integrity mechanisms, when for connections between the home and the care-of ip, as it is likely to traverse the public internet especially in a multi node setup.

**Comment by corhere** [comment]
> moving containers around as needed

Swarm tasks (read: containers managed by Swarm) do not move. They are rescheduled: task A on node A is stopped and removed, then task A' gets started on node B. Task A' is a new container with a distinct identity from its predecessor task A.

> as it is likely to traverse the public internet especially in a multi node setup

Swarm clusters spanning more than one data center is not a supported configuration. Swarm is designed and tuned for all cluster nodes to be on the same LAN.

The only user-facing use case for stable IPv6 addresses for Swarm tasks is for clients to connect directly to an individual task instance, bypassing the [routing mesh](https://docs.docker.com/engine/swarm/ingress/) which means no load-balancing or fault tolerance. The routing mesh takes care of forwarding traffic to the individual tasks, making it the only thing which needs to know how to directly reach the tasks. Since it is integrated into the Swarm control plane, the routing mesh always knows the addresses of every task. Clients only need to know the address of one of the Swarm host nodes to make use of the routing mesh.

Swarm overlay networking and the routing mesh do not yet have support for IPv6, but I am working on it. This feature request sounds more like a partial workaround for this missing functionality in the routing mesh [than a useful feature to have on its own.](https://xyproblem.info/) Are there any use cases (in supported configurations of Swarm) for Mobile IPv6 container addresses that are not better served by a Swarm overlay network and routing mesh which supports IPv6?

**Comment by agowa** [comment]
> Swarm clusters spanning more than one data center is not a supported configuration. Swarm is designed and tuned for all cluster nodes to be on the same LAN

This does not exclude a swarm cluster from spanning more than one data center. A single LAN (you probably mean subnet) can easily span multiple data centers...

**Comment by corhere** [comment]
You seem to be very knowledgeable about IPv6 network architectures, but at the same time are talking about Swarm functionality as if the data center network was the provider. To make sure we are on the same page: Swarm _provides_ multi-tenant network virtualization to containers. Swarm's network architecture implements a subset of the NVO3 architecture described in [RFC 8014](https://datatracker.ietf.org/doc/html/rfc8014) where the containers are the Tenant Systems and each Swarm Node acts as a Network Virtualization Edge (["NVE Co-located with Server Hypervisor"](https://datatracker.ietf.org/doc/html/rfc8014#section-4.1)). Each Swarm Node also acts as an L4 load balancer for Swarm Services. Overlay traffic is encapsulated in VXLAN to provide L2 service to tenant containers over an L3 underlay network.

If I'm understanding you correctly, your proposal is to (ab)use RFC 6275 home addresses as Virtual IPs in order to implement L4 load balancing for IPv6 traffic using Direct-Server Return? Swarm provides L2 overlay networks. _We don't need any support from Layer 3 for DSR!_ The packets can simply be steered to the different backends [by mangling the link-layer destination address](http://www.linuxvirtualserver.org/VS-DRouting.html). It worked with IPv4 and works identically with IPv6.

I'm open to reconsidering how we approach Swarm networking in an IPv6-native world. What would an architecture that takes advantage of RFC 6275 look like, and what would its advantages be over the existing one? It would have to be an architecture which prevents DSR load balancing by mangling the link-layer destination address from working: one which only provides L3 (routed) service to tenant containers. Would Swarm continue to provide virtual network service, or would the nodes themselves be the tenant systems connected to the (virtual) network?

Assuming there is a use-case for RFC 6275 Mobile IPv6 addresses (presumably routable from the Internet) as virtual IPs for load-balanced services, how would that work, e.g. to round-robin load balance per L4 flow? If a client with support for Route Optimization connects to the service's virtual home address, won't its Binding Cache get populated with the care-of address of one particular backend, making the routing "sticky" for that client? Since the routing is at L3, a Binding Update to change the care-of address would take effect immediately. New packets corresponding to existing L4 flows would be instantly rerouted to a different backend container, breaking any established TCP connections. Without Route Optimization the traffic flows are bidirectionally tunneled through the home agent, making it functionally equivalent to traditional NAT load-balancing.

**Comment by agowa** [comment]
> but at the same time are talking about Swarm functionality as if the data center network was the provider

Well, if you design it right and put the right filters for your BGP messages in place you can "unify" it as such and delegate some parts of the logic into the data center.
e.g. with openstack you'd have a dragon speaker into the customer project and the VM within that project would "peer" with what you got from the data center/cloud-provider. And then your workload within these VMs can basically interact with the cloud provider and instruct them to change routing using that peering.

Also it was just one example to show the difference in thinking when designing things with IPv6. We generally try to reduce layers as much as possible without removing the freedom we know from overlays. That's generally where dynamic routing protocols come in and shine.

The other example I mentioned with only router advertisements can be setup even if the cloud provider doesn't offer BGP, however it requires disabling some spoofing protection in the case of a generic openstack deployment at least.


The main thing to consider (and that I'd like to stress on) is that as long as you do NOT utilize these routing capabilities and do not have the actual service IPs within the pods/containers/... you also do NOT get any benefits from IPv6 and you're generally better of with just dual-stacking your Ingress and keeping the cluster single stack (in most cases for compatibility single-stack-IPv4).
Example workload that one'd expect with an IPv6 enabled swarm cluster that would NOT work WITHOUT having End-to-End connections without address-translation are:
* WebRTC (without STUN/TUN)
* Running a WebRTC STUN/TUN within the cluster (without using Node-IPs or raw interfaces [or other hacks])
* Running "ugly" things like e.g. something heavily RPC dependent, like a Samba-AD within the cluster while using a Service-IP for it and not disable redundancy by e.g. bypassing kubeproxy (which technically is also just a fancy-NAT used for load-balancing) and using binding it to a single Node (Node-IP). [Also anything needing a lot of ports is currently an issue with Node-IPs anyway, so e.g. a 1:1 map of 49152-65535, + about 10 well-known and registerd ones)
* Applications that need to know their own IP for usage in Layer 7 protocols.
* FTP-Servers (especially if TLS encrypted as ALGs won't be able to require L7 traffic). If both sides are behind a NAT it just won't work in almost all cases. [and with IPv4 clients most of the time are, thereby currently running a FTP server within a IPv4 swarm cluster isn't possible; And with you also using something similar with IPv6 at least everything in the/another cluster also wouldn't be able to establish FTP connections]
* P2P applications, at least if they want/need to communicate outside of the cluster-too.


> To make sure we are on the same page: Swarm provides multi-tenant network virtualization to containers.

And just fyi, it doesn't need to "layer it's own" in order to accomplish this goal. All the swarm really is, is an interface to setup the multi-tenant networking. Nobody is looking at the wire traffic and says it has to have an additional wrapping around it. We only need it for IPv4, because we do not have any means of dynamically interacting with the upstream. And work around it by masquerading everything from/to the upstream router...

> Overlay traffic is encapsulated in VXLAN to provide L2 service to tenant containers over an L3 underlay network.

You can always add an additional overlay, it was just ONE example architecture (one without the need for a L2 tbh, mainly as it generally only adds pointless complexity. Also I don't know any use case where a L2 within an overlay actually is useful or required at all, even on bare deployments L2 is almost useless these days [except for routing traffic towards the default gateway and for routing/load-balancing protocols themselves to work, ofc]). In general I'd like to question if we should keep L2 "at all costs" or move to a more modern design and have L2 support as an "opt-in" (we can always add another tunnel on-top to allow L2 where needed, and thereby only having to deal design-constraints and issues it causes where it is actually needed [probably never].


> If I'm understanding you correctly, your proposal is to (ab)use RFC 6275 home addresses as Virtual IPs in order to implement L4 load balancing for IPv6 traffic using Direct-Server Return? Swarm provides L2 overlay networks. We don't need any support from Layer 3 for DSR! The packets can simply be steered to the different backends [by mangling the link-layer destination address](http://www.linuxvirtualserver.org/VS-DRouting.html). It worked with IPv4 and works identically with IPv6.

Only if you can mangle the Link-Layer destination or even have one. Many esp. OpenStack deployments are on e.g. IPfabrics which do not have L2 and where it is dropped as soon as it leaves the VM/Node and only the L3 IP is considered by the data-center/cloud-providers overlay... And mainly I'm after E2E visibility and using RFC 6275 is (ab)used here, but not as Virtual IPs but in order to implement Multi-Homing, as almost all containers (last time I checked at least) didn't support multiple interfaces within their container and failed to bind their listeners properly. Using RFC 6275 would work around this by either having both IPs on the same interface [binding to the interface would then bind all of the IPs] or by only exposing the IP Mobility IP towards the containers and thereby having them only know about it [which is kinda like the current overlay, but would allow to unify and optimize traffic flow within the rest of the infrastructure esp. when using service (anycast/non-unique) IPs. To be fair, this consideration also already includes a lot of traffic-engineering concerns.
(I don't like the term "virtual-ip", as it is way too ambiguous. First of all almost all IPs are virtual [as in within/on a hypervisor of some sort] and second some times the "virtual-ip" is configured on the upstream interface and visible within ARP/ND-proxy like with VRRP, and other times it isn't like with e.g. BGP and DHCP-PD where it is on a loop-back interface. Other times "virtual-ip" means one that is NAT-ed onto and not configured on any of the interfaces directly [but simultaneously in this very same setup this IP is also often NOT considered "virtual-ip" but call the "public-ip" of that service...] In general this term is not just not-useful in communicating ideas, but because of different people associating different things with it [and all having different limitations and implications] more of an obstacle. But if you mean by "virtual-ip" one that the application itself would bind its listening socket to like within the loop-back interface case then yes. But if you're thinking of one of the others then no. The difference being, the service IP would stay on the interface as long as the namespace/pod is running [with the VRRP one it would get de-configured on the passive node and only be configured on the interface of the active node] and the application would see that IP when querying the interface IPs [which in case of being on the passive node or with a dedicated load-balancer holding the virtual-ip it won't, therefore it also wouldn't open it's listening socket on it].)


> Would Swarm continue to provide virtual network service, or would the nodes themselves be the tenant systems connected to the (virtual) network?

Tl;Dr: E2E visibility and improved efficiency in traffic routing (thereby more throughput and less resource utilization). The swarm itself wouldn't be within the traffic flow anymore but it would still control/manage it.

I already outlined most of the differences as well as benefits above, but in other words Swarm in this case would mainly be an API used to manage, abstract, and filter communication upstream towards the cloud provider (as well as the one required between nodes, for full mesh routing without additional overlay, e.g. when using BGP it would setup everything needed for signed BGP updates flowing between the nodes and the nodes updating their routing tables accordingly [last part is done by a BGP daemon, but the swarm would bootstrap it in this example]) in order to provide multi-tenancy within a single tenant of theirs. E.g. Swarm would reject a request of one of its tenants for a service IP it hasn't been allocated/configured for. It also would allow lower-privileged work load to dynamically change the infrastructure which normally requires root privileges, like get a new service-ip/prefix or remove/join itself from/to its service-ip/prefix.

> Assuming there is a use-case for RFC 6275 Mobile IPv6 addresses (presumably routable from the Internet) as virtual IPs for load-balanced services, how would that work, e.g. to round-robin load balance per L4 flow? If a client with support for Route Optimization connects to the service's virtual home address, won't its Binding Cache get populated with the care-of address of one particular backend, making the routing "sticky" for that client? Since the routing is at L3, a Binding Update to change the care-of address would take effect immediately. New packets corresponding to existing L4 flows would be instantly rerouted to a different backend container, breaking any established TCP connections. Without Route Optimization the traffic flows are bidirectionally tunneled through the home agent, making it functionally equivalent to traditional NAT load-balancing.

In short ECMP routers support L3/L4 stickiness.
However in the case that feature should be absent or we've a non ECMP upstream router to work with we could use a DSR-HAProxy (or even a VM with IPtables/eBGF if session synchronization is worked out at least) as it is basically a ECMP router then. Well, for the above example it would need a bit of glue code to dynamically populate backends using the next_hop information received through BGP and towards upstream all either one or all (depending on what the provider supports, aka. if active/active or active/passive can be used) of these DSR-HAProxy nodes is advertised for all of the prefixes (Same for load-balancer appliances).


I'm no longer at the company I was when I opened that ticket, so I don't have access to my notes from back then anymore. Most of of what I said so far is from the top of my head trying to think what we tried to do back then. I think I mixed up some parts of the "justification".
These days I'd probably introduce and focus the justification for adding native RFC6275 support more around, multi-homing (with upstream source filtering) and renumbering support. As well as roaming between networks where different address allocation mechanisms are used. Like one may use DHCPv6-PD, one may only offer ND-Proxy, and another one may not have either. As such issues are common on SOHO and SMB deployments, as well as all of the local "testing" clusters running on notebooks. (The roaming one would be mostly notebooks, and the main concern would be keeping outbound traffic functioning, as inbound also wouldn't work with IPv4 in similar situations...)

Also by now I'd probably in addition suggest an alternative (but way more drastic) design using **CGAs+HIPv2+SEND+ECMP**
All of the IPs (pod, node, services, ...) are Cryptographically Generated Address (CGAs). For spoofing protection, have them non consecutively allocated, as prerequisite for SEND. SEND (Secure Neighbor Discovery for the setting lifetime and validity to 0 that normal Neighbor Discovery doesn't allow for it because of not mentioned and questionable "security" reasons). HIP(v2) for separating the end-point identifier and locator roles of IPs, to enable renumbering and multi-homing without requiring application support as their bindings still wouldn't change. RFC6275 is basically a subset of HIPv2 with a slightly different proposed implementation.


If I find time I'll try to write a blog post with a kind of mix of a post mortem and design doc with what I still recall from back then, but I'll be way too much for this ticket, as it probably spans multiple pages but if you'd like I'll link it then. As it may help understand where the request for RFC6275 is coming from. It was the result of hitting multiple and different design limitations that all boiled down to the absence of true E2E within swarm (or k8s).

**Comment by agowa** [comment]
As my previous reply was way too long I'll try to summarize it and then hide the former post.

* ECMP routers would provide L3/L4 stickyness.
* If no L3/L4 stickiness is available (or the load balancing algorithm is not desirable) use something like HAProxy in DSR mode. However that requires a bit of glue clode to have next_hops from BGP as backends in HAProxy and work with other limitations like the cloud provider only allowing a single next hop and no BGP. So that glue code would also have to call the cloud provider API to update a static route to another node.
* Issues with multi-homing in SOHO and SMB where no allocated IPv6-PI is available, the ISP doesn't speak BGP and only hands down a static prefix with source address filtering. The network would be dual-addressed using both prefixes simultaneously. RFC6275 would keep outbound connections working without having to restart/redeploy containers. The socket bindings would stay valid and all pods would still be locally be able to make outbound connections as well as towards to service IPs themselves.
* On notebooks we've issues with roaming to different networks, we're only concerned about outbound connectivity as well as from the host itself. With RFC6275 support we also get this. Inbound connectivity from public internet wouldn't work in these cases anyway, then basically being the same as with IPv4-only clusters on notebooks.
* The overlay would become "virtual" aka only a design thing and not be visible within the package headers.
* Docker swarm would become basically an API Endpoint for interacting with the underlay routing in a controlled and manageable way.
* The kind of convergence of the overlay into the underlay is similar to what kube-router does. Aka no VXLAN and we address what we want.
* This ticket is for RFC6275 support, so the load balancing part is kinda out of scope for it. In the BGP example it would be through the ECMP routers supporting L4 stickynes for example. (Or a DSR-HAProxy being inserted, or small eBPF hook, ...)
* These days instead of "only" asking for RFC6275 I'd probably also suggest the entire CNI stack to be redesigned using CGAs+HIPv2+SEND+ECMP. However only RFC6275 alone would be enough to be able to work around most of the issues. Also RFC6275 and HIPv2 are basically the same idea with a slightly different twist on the suggested implementation. They're both kinda abstract outlines. RFC6275 is smaller where as the combination of the other with HIPv2 having extensions that when also implemented would provide way more flexibility (e.g. in regards to load balancing and multi homing) that otherwise need to be diy-ed in.
* L2 within the overlay is unecessary. There are no usecases for it. Dropping it and also having E2E visibility would allow to remove all overlays and unify down to the cloud providers underlay. However in practice we probably will stay within the cloud providers tenant network and only drop the VXLAN from within the swarm.
* I'll try to write a (or multiple) blog posts later trying to outline all of the history that went into this request. Even me trying to shorten it before resulted in that wall of text. And this bullet point summary is also already way too long.

**Comment by corhere** [comment]
> And mainly I'm after E2E visibility and using RFC 6275 is (ab)used here, but not as Virtual IPs but in order to implement Multi-Homing, as almost all containers (last time I checked at least) didn't support multiple interfaces within their container and failed to bind their listeners properly.

Linux supports binding multiple IP addresses (v4 and v6) to any interface. Docker (clunkily) makes it possible to bind additional addresses using the `docker network connect --link-local-ip` flag. Connecting a container to more than one network results in the presence of multiple interfaces within the container. Containers connected to a Swarm overlay network implicitly get connected to the `docker_gwbridge` network through a second interface through which (NAT'ed IPv4) north-south connectivity is provided.

In the common case, all a container process has to do to support multi-homing is to listen on the unspecified address.

> * Issues with multi-homing in SOHO and SMB where no allocated IPv6-PI is available, the ISP doesn't speak BGP and only hands down a static prefix with source address filtering. The network would be dual-addressed using both prefixes simultaneously. RFC6275 would keep outbound connections working without having to restart/redeploy containers. The socket bindings would stay valid and all pods would still be locally reachable, as well als able to make outbound connections, including towards all of the service IPs.

Why would _Mobile_ IPv6 be necessary if the prefix is static, given that no renumbering would ever happen? What sorts of events would otherwise require containers to be restarted or redeployed if RFC6725 isn't implemented?

> * On notebooks we've issues with roaming to different networks

Are we still talking about Swarm here? Are you running a single-node Swarm on one developer laptop? Are you building a cluster out of multiple laptops which roam independently??!

**Comment by agowa** [comment]
> Are you running a single-node Swarm on one developer laptop?

Yes

> Why would Mobile IPv6 be necessary if the prefix is static

It was kinda a hack to get better application support for the multi-interface and multi-ip case. As well as have non NATed IPs.
I had it in the 4000 words I cut from my original answer. When recalling what we did back then it was also based on multi-homing (when a network has multiple uplinks from different ISPs and both links do source filtering and everything is dual-addressed. Aka has one IPv6 per uplink).

In short it was multi-factual and the main thing was to split the two roles of IPs to have more flexibility within the containers. And as I said these days I'd probably suggest to use HIPv2 within the design, as it is less of a hack but only shines when implemented and used correctly. 

> Linux supports binding multiple IP addresses (v4 and v6) to any interface. 

Yes, Linux can. However at least back then it didn't work with containers, mainly because they were built in a way that they didn't expect and handle them well.

> Docker (clunkily) makes it possible to bind additional addresses using the docker network connect --link-local-ip flag

That doesn't count as docker supporting it. But at least the cni has been improved(?) in that regards? I don't have the ticket id from back then when we reported/asked about these challenges nor do I recall all of the details anymore but it was said to be "a design limitation" of the cni and closed as won't fix...

> Connecting a container to more than one network results in the presence of multiple interfaces within the container

Try it with common containers. At least back then most apps only bound to one interface. Which one was also not deterministic. And some failed to start entirely.

> through a second interface through which (NAT'ed IPv4) north-south connectivity is provided.

Well, the issue is when you don't have the NAT. Then it doesn't work that well... (Also true for IPv4, but way less common for people to have enough IPv4s for this to actually surface).
